packages/vite/src/node/build.ts

```diff
@@ -875,13 +889,66 @@ async function buildEnvironment(
       prepareOutDir(resolvedOutDirs, emptyOutDir, environment)
     }
 
-    const res: RolldownOutput[] = []
-    for (const output of normalizedOutputs) {
-      res.push(await bundle[options.write ? 'write' : 'generate'](output))
-    }
+    const res = await build()
+
     logger.info(
       `${colors.green(`✓ built in ${displayTime(Date.now() - startTime)}`)}`,
     )
+
+    async function build() {
+      const res: RolldownOutput[] = []
+      for (const output of normalizedOutputs) {
+        // TODO(underfin): using the generate at development build could be improve performance.
+        res.push(await bundle![options.write ? 'write' : 'generate'](output))
+      }
+
+      if (server) {
+        // watching the files
+        for (const file of bundle!.watchFiles) {
+          if (path.isAbsolute(file) && fs.existsSync(file)) {
+            server.watcher.add(file)
+          }
+        }
+
+        // Write the output files to memory
+        for (const output of res) {
+          for (const outputFile of output.output) {
+            server.memoryFiles[outputFile.fileName] =
+              outputFile.type === 'chunk' ? outputFile.code : outputFile.source
+          }
+        }
+      }
+      return res
+    }
+
+    if (server) {
+      server.watcher.on('change', async (file) => {
+        const startTime = Date.now()
+        const hmrOutput = (await bundle!.generateHmrPatch([file]))!
+        if (hmrOutput.patch) {
+          const url = `${Date.now()}.js`
+          server.memoryFiles[url] = hmrOutput.patch
+          server.ws.send({
```

Comment:
It's fine for now, but we should add `environment.hot.send` and use that in the future to make this compatible with environment API.

Reply:
I'm not similar to environment API, it is confused for me... We need to consider the pr effect the environment API.